### PR TITLE
fuzz: introduce more fuzzers

### PR DESCRIPTION
**Pull request by mrc0mmand**
I finally got some time to play around with fuzzers and attempted to write a few dead-simple ones. I'm not sure if they're going to be useful at all, but I think I'm slowly moving from the *"I have absolutely no idea what I'm doing*" phase to the *"oh, this might actually work"* one. Feedback is definitely more than welcome.

The first three fuzzers haven't found anything so far, but the `nspawn-settings` one caught a (possible) memory leak:

```
==14==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 64 byte(s) in 1 object(s) allocated from:
    #0 0x4f28fe in realloc /src/llvm/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:164
    #1 0x7f38101b8efb in greedy_realloc /work/build/../../src/systemd/src/basic/alloc-util.c:55:13
    #2 0x7f38101f4c13 in extract_first_word /work/build/../../src/systemd/src/basic/extract-word.c:46:22
    #3 0x7f38101f5f6a in extract_many_words /work/build/../../src/systemd/src/basic/extract-word.c:257:21
    #4 0x53fb32 in bind_mount_parse /work/build/../../src/systemd/src/nspawn/nspawn-mount.c:192:13
    #5 0x539a57 in config_parse_bind /work/build/../../src/systemd/src/nspawn/nspawn-settings.c:281:13
    #6 0x7f381011693e in next_assignment /work/build/../../src/systemd/src/shared/conf-parser.c:132:32
    #7 0x7f3810111ca5 in parse_line /work/build/../../src/systemd/src/shared/conf-parser.c:257:16
    #8 0x7f3810110bd7 in config_parse /work/build/../../src/systemd/src/shared/conf-parser.c:377:21
    #9 0x538167 in settings_load /work/build/../../src/systemd/src/nspawn/nspawn-settings.c:48:13
    #10 0x537c38 in LLVMFuzzerTestOneInput /work/build/../../src/systemd/src/fuzz/fuzz-nspawn-settings.c:25:16
    #11 0x58dc66 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/libfuzzer/FuzzerLoop.cpp:529:15
    #12 0x58c399 in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool*) /src/libfuzzer/FuzzerLoop.cpp:453:3
    #13 0x5910b4 in fuzzer::Fuzzer::ReadAndExecuteSeedCorpora(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, fuzzer::fuzzer_allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, fuzzer::fuzzer_allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&) /src/libfuzzer/FuzzerLoop.cpp:755:7
    #14 0x592278 in fuzzer::Fuzzer::Loop(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, fuzzer::fuzzer_allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, fuzzer::fuzzer_allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&) /src/libfuzzer/FuzzerLoop.cpp:779:3
    #15 0x55aa5f in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/libfuzzer/FuzzerDriver.cpp:776:6
    #16 0x54d62c in main /src/libfuzzer/FuzzerMain.cpp:19:10
    #17 0x7f380ef4b82f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)

SUMMARY: AddressSanitizer: 64 byte(s) leaked in 1 allocation(s).

```

Reproducer can be found under `test/fuzz/fuzz-nspawn-settings/leak-4ff0e2498f596a77ea68d185c61e9e9ff9bb657f`.

### CODE REVIEW

File: src/fuzz/meson.build [hunk 1]
```diff
@@ -116,4 +116,17 @@ fuzzers += [
...
+         []],
+
+        [['src/fuzz/fuzz-hostname-util.c'],
+         [libshared],
+         []],
+
+        [['src/fuzz/fuzz-nspawn-settings.c'],
+         [libshared,
+          libnspawn_core],
+         []],
```
Comment (keszybz) [inline src/fuzz/meson.build:131]: I'd prefer to have this split among the three commits that add those files, like it is done with the first commit.

**Comment by keszybz** [comment]
I pushed a fix for the memleak.

Let's wait for @titanous's review too though.

**Comment by mrc0mmand** [comment]
Thanks for the fix! I'm definitely for dropping the more-or-less useless fuzzers, as they indeed cover a minuscule part of the codebase. I was looking for *easy-to-fuzz* functions to see how the machinery works, so they might be replaced in the future by a better fuzzer which covers a significantly larger portion of the codebase.

**Comment by titanous** [comment]
I don't think there's any problem with small fuzzers, as long as we aren't adding hundreds of them. Google has cores to spare 😂

**Comment by poettering** [comment]
(do we already fuzz calendar and timestamp expressions? sounds like something easy too)

**Comment by mrc0mmand** [comment]
> 
> 
> (do we already fuzz calendar and timestamp expressions? sounds like something easy too)

Will look into that!

**Pull request merged**